### PR TITLE
Migrating Get/Grant/Revoke/Test NTFS permissions from Carbon

### PR DESCRIPTION
**Pull request by splatteredbits**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,49 @@
-# 1.0.1
+<!--markdownlint-disable MD012 no-multiple-blanks-->
 
-* Adding Guid to `Carbon.FileSystem.psd1` which is needed to publish to PSGallery.
+# Carbon.FileSystem Changelog
 
-# 1.0.0
+## 1.1.0
```
Comment (decoyjoe) [inline CHANGELOG.md:5]: Add release date.

File: Carbon.FileSystem/Carbon.FileSystem.psd1 [hunk 2]
```diff
@@ -78,7 +78,11 @@
 
     # Functions to export from this module. Only list public function here.
     FunctionsToExport = @(
-        'Get-FileHardLink'
+        'Get-CNtfsPermission',
+        'Get-FileHardLink',
```
Comment (decoyjoe) [inline Carbon.FileSystem/Carbon.FileSystem.psd1:83]: Why does `Get-FileHardLink` not contain the `CNtfs` prefix?

File: Carbon.FileSystem/Carbon.FileSystem.psm1 [hunk 3]
```diff
@@ -16,10 +20,12 @@
 Set-StrictMode -Version 'Latest'
 
 # Functions should use $moduleRoot as the relative root from which to find
-# things. A published module has its function appended to this file, while a 
+# things. A published module has its function appended to this file, while a
 # module in development has its functions in the Functions directory.
 $moduleRoot = $PSScriptRoot
 
+Import-Module -Name (Join-Path -Path $PSScriptRoot -ChildPath 'Modules\Carbon.Permissions' -Resolve)
```
Comment (decoyjoe) [inline Carbon.FileSystem/Carbon.FileSystem.psm1]: Disable verbose module import statements, add `-Verbose:$false`.

File: Carbon.FileSystem/Functions/Get-CNtfsPermission.ps1 [hunk 4]
```diff
@@ -0,0 +1,70 @@
+
+function Get-CNtfsPermission
+{
+    <#
+    .SYNOPSIS
+    Gets the permissions (access control rules) for a file or directory.
+
+    .DESCRIPTION
+    Permissions for a specific identity can also be returned. Access control entries are for a path's discretionary
```
Comment (decoyjoe) [inline Carbon.FileSystem/Functions/Get-CNtfsPermission.ps1]: What happens if there are no permissions for that identity? An error? Nothing?

File: Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1 [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    `Path` parameter, the user/group name to the `Identity` parameter, and the permissions to the `Permission`
+    parameter. By default, the permissions are applied to the folder and inherited to all its subfolders and files. To
+    control how the permissions are applied, use the `ApplyTo` parameter. If you want permissions to only apply to child
+    files and folders, use the `OnlyApplyToChildFilesAndFolders` switch.
+
+    By default, an "Allow" permission is granted. To add a "Deny" permission, set the value of the `Type` parameter to
+    `Deny`.
+
+    All existing, non-inherited permissions for the given identity are removed first. If you want to preserve a
+    user/group's existing permissions, use the `Append` switch.
```
Comment (decoyjoe) [inline Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1:18]: Seems backwards to me. `Grant` to me implies append.

I'd opt for `Set` as the verb if the function behaves in this way.

File: Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1 [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    To remove *all* non-inherited permissions except the permission being granted, use the `Clear` switch.
+
+    The permission is only granted if it doesn't exist. To always grant the permission, use the `Force` switch.
```
Comment (decoyjoe) [inline Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1:22]: I'd rephrase from `To always grant the permission` to `To always apply the permission` since the function can add Deny permissions (i.e. not a "grant").

File: Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1 [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    .EXAMPLE
+    Grant-CNtfsPermission -Identity ENTERPRISE\Engineers -Permission FullControl -Path C:\EngineRoom -Clear
+
+    Grants the Enterprise's engineering group full control on the engine room.  Any non-inherited, existing access rules
+    are removed from `C:\EngineRoom`.
+
+    .EXAMPLE
+    Grant-CNtfsPermission -Identity BORG\Locutus -Permission FullControl -Path 'C:\EngineRoom' -Type Deny
+
+    Demonstrates how to grant deny permissions on an objecy with the `Type` parameter.
```
Comment (decoyjoe) [inline Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1:60]: "grant deny" sounds off.

File: Carbon.FileSystem/Functions/Test-CNtfsPermission.ps1 [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    .SYNOPSIS
+    Tests if permissions are set on a folder or file.
+
+    .DESCRIPTION
+    The `Test-CNtfsPermission` function tests if an identity has a permission on a file/folder. Pass the path to check
+    to the `Path` parameter, the user/group name to the `Identity` parameter, and the permission to check for to the
+    `Permission` parameter. If the user/group has the given permission on the given path, the function returns `$true`,
+    otherwise it returns `$false`.
+
+    Inherited permissions on *not* checked by default.  To check inherited permission, use the `-Inherited` switch.
```
Comment (decoyjoe) [inline Carbon.FileSystem/Functions/Test-CNtfsPermission.ps1]: Typo: `on *not*`

File: Carbon.FileSystem/Functions/Test-CNtfsPermission.ps1 [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    param(
+        # The path to a folder/file on which the permissions should be checked.
+        [Parameter(Mandatory)]
+        [String] $Path,
+
+        # The user or group name whose permissions to check.
+        [Parameter(Mandatory)]
+        [String] $Identity,
+
+        # The permission to test for: e.g. FullControl, Read, etc.
```
Comment (decoyjoe) [inline Carbon.FileSystem/Functions/Test-CNtfsPermission.ps1]: Link to or document valid values?

File: Carbon.FileSystem/prism.json [hunk 10]
```diff
@@ -0,0 +1,10 @@
+{
+	"PSModules": [
+    {
+      "Name": "Carbon.Permissions",
+      "Version": "1.*-*"
+    }
```
Comment (decoyjoe) [inline Carbon.FileSystem/prism.json:6]: Inconsistent indentation.

File: prism.json [hunk 11]
```diff
@@ -0,0 +1,12 @@
+{
+	"PSModules": [
+    {
+      "Name": "Carbon",
+      "Version": "*"
+    },
+    {
+      "Name": "Carbon.Accounts",
+      "Version": "1.*-*"
+    }
```
Comment (decoyjoe) [inline prism.json:10]: Inconsistent indentation.

**Review by decoyjoe — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (splatteredbits) [reply CHANGELOG.md:5]: No release date yet. This will stay as a prerelease version until then.

File: Carbon.FileSystem/Carbon.FileSystem.psd1 (hunk 2 above)
Comment (splatteredbits) [reply Carbon.FileSystem/Carbon.FileSystem.psd1:83]: Will rename.

Comment (splatteredbits) [reply Carbon.FileSystem/Carbon.FileSystem.psd1:83]: It was like this when I got here.

File: Carbon.FileSystem/Functions/Get-CNtfsPermission.ps1 (hunk 4 above)
Comment (splatteredbits) [reply Carbon.FileSystem/Functions/Get-CNtfsPermission.ps1]: Updated docs to explain. Currently, no error is written if the identity doesn't have permissions.

File: Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1 (hunk 7 above)
Comment (splatteredbits) [reply Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1:60]: That's the way NTFS permissions are designed: there are allow rules and deny rules, but they both get granted the same way. It's a permission type, not a permission.

File: Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1 (hunk 6 above)
Comment (splatteredbits) [reply Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1:22]: That's the way NTFS permissions are designed: there are allow rules and deny rules, but they both get granted the same way. It's a permission type, not a permission.

File: Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1 (hunk 5 above)
Comment (splatteredbits) [reply Carbon.FileSystem/Functions/Grant-CNtfsPermission.ps1:18]: This function cares about two parts of an NTFS access rule: an identity and the permission . The underlying object, an access rule, actually has four parts: the permission, its type (allow/deny), the identity, and how to apply the permission to ancestors. When our automation grants permissions, 99% of the time, that's all we care about: a specific identity gets/nee. I don't even think the ability to append extra access rules came from us. It might have been a PR.

Your comment got me thinking, and maybe what we should also have are `Add-CNtfsAccessRule`, `Remove-CNtfsAccessRule`, and `Set-CNtfsAccessRule` functions that manage the actual access rules and allow more advanced scenarios. I'll keep thinking about this.
